ql/src/java/org/apache/hadoop/hive/ql/queryhistory/persist/AbstractQueryHistoryPersistor.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.queryhistory.persist;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.TableType;
+import org.apache.hadoop.hive.metastore.api.Database;
+import org.apache.hadoop.hive.metastore.api.FieldSchema;
+import org.apache.hadoop.hive.metastore.api.hive_metastoreConstants;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.apache.hadoop.hive.ql.parse.PartitionTransform;
+import org.apache.hadoop.hive.ql.parse.TransformSpec;
+import org.apache.hadoop.hive.ql.queryhistory.schema.QueryHistorySchema;
+import org.apache.hadoop.hive.ql.session.SessionStateUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public abstract class AbstractQueryHistoryPersistor implements QueryHistoryPersistor {
+  protected Logger LOG = LoggerFactory.getLogger(getClass());
+  @VisibleForTesting
+  HiveConf conf;
+  protected QueryHistorySchema schema;
+
+  public void init(HiveConf conf, QueryHistorySchema schema) {
+    this.conf = conf;
+    this.schema = schema;
+
+    Database database = initDatabase();
+    Table table = initTable(database);
+    try {
+      postInitTable(table);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  protected Database initDatabase() {
+    try (Hive hive = Hive.get(conf)) {
```

Comment:
`Hive.get(conf)` gets hive object for the current thread. If one is not initialized then a new one is created
It's an expensive operation, so doesn't make sense to open & close in initDB and initTable

Reply:
okay, let me pass this around the different usages of the init phase